### ISSUE TITLE
G(tk)Application support in gtk3reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5492]
|<img alt="lucian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| lucian reported|
|-|-|
|Trac ID|trac#5492|
|Type|enhancement|
|Created|2012-02-19 20:59:26Z|
|Branch|https://github.com/twisted/twisted/tree/gtk-app-5492|

GLib has a GApplication class that handles things like application uniqueness and a few other things, and provides its own .run() method. GtkApplication inherits from it and also does Gtk-specific things.

After installing the gtk3reactor, if I call reactor.run() the UI does show up, but the GApplication's 'activate' event never fires. At glyph's suggestion I explicitly set the reactor's mainloop to that of the GApplication's and 'activate' fired as expected.

The reactor also doesn't stop when I close the window, even with the hack.

Attachments:

* [play.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/1409f59a4361208f27b50c7e9a4206cf3388fc89/4ac4063ceead212fffd0bbe33cc7bd1434d4fd31.py) (552 bytes) - added by lucian on 2012-02-20 11:07:57Z - Example of hanging reactor.
* [play.2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/1409f59a4361208f27b50c7e9a4206cf3388fc89/0236968b5229bf467844dfd464b21fcb4e97503e.py) (710 bytes) - added by lucian on 2012-02-20 11:14:46Z - Example with Window and a twisted callback
* [tx_gapp.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/1409f59a4361208f27b50c7e9a4206cf3388fc89/1ea19245cfc5ec7d00a27d6c674969664548c95a.patch) (1592 bytes) - added by lucian on 2012-02-20 19:09:05Z - Patch to replicate the _run=app.run and installSignalHandlersFalse=False workarounds
* [tx_gapp2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/1409f59a4361208f27b50c7e9a4206cf3388fc89/ca26e3b7fc75b574d105379001c359845c2b97ec.patch) (2608 bytes) - added by lucian on 2012-02-20 20:18:39Z - Default back to installing signal handlers by default, at exarkun's suggestion. Add docs.
* [tx_gapp3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/1409f59a4361208f27b50c7e9a4206cf3388fc89/a8e2e274e69dc8c23d4a6012b325fc2488bad144.patch) (4727 bytes) - added by lucian on 2012-02-20 21:47:00Z - Remove a useless method, add tests.

<details><summary>Searchable metadata</summary>

```
trac-id__5492 5492
type__enhancement enhancement
reporter__lucian lucian
priority__normal normal
milestone__ 
branch__branches_gtk_app_5492 branches/gtk-app-5492
branch_author__lucian lucian
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1329685166000000 1329685166000000
changetime__1336862670000000 1336862670000000
version__None None
owner__itamar itamar
cc__dobey
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

Could you provide a minimal example?

**Comment by twisted-trac** [comment]
|<img alt="lucian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|lucian commented|
|-|-|

Perhaps something like the reactor.registerWxApp would work?

http://twistedmatrix.com/documents/current/core/howto/choosing-reactor.html#auto12

**Comment by twisted-trac** [comment]
|<img alt="lucian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|lucian commented|
|-|-|

The sedonc minimal example doesn't hang when I close the window. I believe it happened because of my usage of IRCClient in my app.

I was wrong, I should've written the minimal example before filing the ticket.

**Comment by twisted-trac** [comment]
|<img alt="lucian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|lucian commented|
|-|-|

If I start the reactor with .run(installSignalHandlers=False), the app responds to SIGTERM and dies correctly, and also responds to SIGINT (incorrectly, because of a GTK/py-gi bug https://bugzilla.gnome.org/show_bug.cgi?id=622084)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

If possible the tests should use [[#5547](https://github.com/twisted/twisted/issues/5547)](https://github.com/twisted/twisted/issues/5547). If #5547 isn't ready and we want to release 12.1, we should at a minimum make sure we do the SIGCHLD cleanup stuff that `ReactorBuilder` does since these tests create a new reactor instance and run it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

(In [34040]) Branching to 'gtk-app-5492'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to @itamarst|
|-|-|

I made some fixes in the branch, in particular matching coding standard a bit better, and moving the code to gireactor instead of _glibbase.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst removed owner|
|-|-|

OK, ready for final review assuming all tests are happy:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gtk-app-5492

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @itamarst|
|-|-|

 * I'm seeing test failures when using the gi and gtk3 reactor on the whole test suite. They got attached to random tests:

```

[ERROR]
Traceback (most recent call last):
  File "Twisted/trunk/twisted/internet/base.py", line 800, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "Twisted/trunk/twisted/internet/_glibbase.py", line 296, in stop
    posixbase.PosixReactorBase.stop(self)
  File "Twisted/trunk/twisted/internet/base.py", line 577, in stop
    "Can't stop reactor that isn't running.")
twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running.
```
 I think maybe overriding `_run` may not be enough?

 * choosing-reactor.xhtml has some tabs in it. It'd be nice to wrap lines to 80 chars too.

Do we have a gtk3 slave somewhere too?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

So in theory I fixed it. But I can't test it since it's not possible to do this correctly on older versions of gobject introspection, AFAICT. Removing from 12.1 milestone, though, this isn't really necessary for basic gtk3 support; if we make it in then fine, but doesn't seem like it should be a blocker.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

Next step: run tests on e.g. Ubuntu 12.04. I will upgrade my system from 11.10 real soon now, but feel free to run on your system and if it looks good it's mergeable in theory.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

Looks good, tests are passing for me now. I think we can merge it as it is. Maybe the last question I have is whether we should raise an error when registerGApplication is called after reactor.run or called a second time.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst commented|
|-|-|

OK, I added those two things, will merge once buildbot is happy.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst set status to `closed`|
|-|-|

(In [34391]) Merge gtk-app-5492: Add support for Gio.Application instance registration.

Fixes: [#5492](https://github.com/twisted/twisted/issues/5492)
Author: lucian, itamar
Review: itamar, therve
